zcash_client_backend/src/fees/zip317.rs

```diff
@@ -67,129 +71,157 @@ impl ChangeStrategy for SingleOutputChangeStrategy {
-                // effects at this layer.
-                let t_grace_dust = available_grace_inputs.saturating_sub(t_disallowed_dust);
-                t_disallowed_dust = t_disallowed_dust.saturating_sub(t_grace_dust);
-
-                let s_grace_dust = available_grace_inputs
-                    .saturating_sub(t_grace_dust)
-                    .saturating_sub(s_disallowed_dust);
-                s_disallowed_dust = s_disallowed_dust.saturating_sub(s_grace_dust);
-
-                let o_grace_dust = available_grace_inputs
-                    .saturating_sub(t_grace_dust)
-                    .saturating_sub(s_grace_dust)
-                    .saturating_sub(o_disallowed_dust);
-                o_disallowed_dust = o_disallowed_dust.saturating_sub(o_grace_dust);
-            }
-
-            // Truncate the lists of inputs to be disregarded in input selection to just the
-            // disallowed lengths. This has the effect of prioritizing inputs for inclusion by the
-            // order of the original input slices, with the most preferred inputs first.
-            transparent_dust.reverse();
-            transparent_dust.truncate(t_disallowed_dust);
-            sapling_dust.reverse();
-            sapling_dust.truncate(s_disallowed_dust);
-            orchard_dust.reverse();
-            orchard_dust.truncate(o_disallowed_dust);
+            let mut orchard_dust: Vec<NoteRefT> = vec![];
 
+            // Depending on the shape of the transaction, we may be able to spend up to
+            // `grace_actions - 1` dust inputs. If we don't have any dust inputs though,
+            // we don't need to worry about any of that.
             if !(transparent_dust.is_empty() && sapling_dust.is_empty() && orchard_dust.is_empty())
             {
-                return Err(ChangeError::DustInputs {
-                    transparent: transparent_dust,
-                    sapling: sapling_dust,
+                let t_non_dust = transparent_inputs.len() - transparent_dust.len();
+                let t_allowed_dust = transparent_outputs.len().saturating_sub(t_non_dust);
+
+                // We add one to either the Sapling or Orchard outputs for the (single)
+                // change output. Note that this means that wallet-internal shielding
+                // transactions are an opportunity to spend a dust note.
+                let net_flows = calculate_net_flows::<NoteRefT, Self::FeeRule, Self::Error>(
+                    transparent_inputs,
+                    transparent_outputs,
+                    sapling,
                     #[cfg(feature = "orchard")]
-                    orchard: orchard_dust,
-                });
+                    orchard,
+                    #[cfg(feature = "transparent-inputs")]
+                    ephemeral_parameters.ephemeral_input_amount(),
+                    #[cfg(feature = "transparent-inputs")]
+                    ephemeral_parameters.ephemeral_output_amount(),
```

Comment:
We checked above that the ephemeral parameters are the default, so these are both `None`.